Tutorial for optimization

TiKeil opened the pull request:
This PR contains a very detailed tutorial for optimization. The content is highly related to my talk on the online pymor school 2020. 
I will rebase this PR after #1110  has been merged. So please ignore all commits except the last two and only look at [this file](https://github.com/TiKeil/pymor/blob/tutorial_optimization/docs/source/tutorial_optimization.rst).

@pymor/all it would be nice if you could read through it, there might be some typos or other issues. 

Let me know what you think.

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+.. math::
+
+    
+   \min_{\mu \in \mathcal{P}} \mathcal{J}(\mu),  \tag{$\hat{P}$}
```
```suggestion
   \min_{\mu \in \mathcal{P}} \mathcal{J}(\mu).  \tag{$\hat{P}$}
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+using MOR methods. Some of them rely on an RB method with traditional
+offline/online splitting, which typically result in a very online
+efficient approach. Recent research also tackles overall efficiency by
+overcoming the expensive offline approach which we will discuss further
```
```suggestion
overcoming the expensive offline phase, which we will discuss further
```

renefritze commented on src/pymor/models/interface.py:
```diff
@@ -179,9 +255,9 @@ def _compute_output_error_estimate(self, solution, mu=None, **kwargs):
-    def compute(self, solution=False, output=False,
-                solution_error_estimate=False, output_error_estimate=False, *,
-                mu=None, **kwargs):
+    def compute(self, solution=False, output=False, solution_d_mu=False, output_d_mu=False,
```
Shouldn't the new args come after the "old" ones for compat with positional usage?

TiKeil replied:
It is true that is does not preserve positional usage but at the same time I found the order of the arguments way better than just appending them to the current state. IMO, the arguments for the estimates should come after the others...
The `compute` method has anyway also been introduced in the upcoming release which made me think it is okay to still restructure arguments a bit.

renefritze replied:
The recent introduction of the compute method is indeed a very good argument. Aside from that, I have no strong opinion wrt estimates before others or not. 
That also somewhat means we really should get this PR into 2020.2 though.

TiKeil replied:
I agree. I also do not really have the strongest opinion about the order of the arguments.

@sdrave , do you have any particular preference?

renefritze commented on src/pymortests/demos.py:
```diff
@@ -44,6 +44,7 @@
     ('burgers', ['--num-flux=lax_friedrichs', '0.1']),
     ('burgers', ['--num-flux=engquist_osher', '0.1']),
     ('burgers', ['--num-flux=simplified_engquist_osher', '0.1']),
+    ('linear_optimization', [40, 20]),
```
Can you comment on the runtime impact in CI? Relative to other demos?

TiKeil replied:
The runtime is about 2 seconds atm. I reduced this quite a bit.

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+    output_functional = fom.rhs.H * theta_J 
+    fom = fom.with_(output_functional=output_functional)
+
+To overcome that pyMORs outputs return a \|NumpyVectorArray\|, we have
```
Wasn't this changed in #1089?

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+.. math::
+
+   \begin{align}
+   l(x, y) &= \tfrac{1}{2} \pi^2 cos(\tfrac{1}{2} \pi x) cos(\tfrac{1}{2} \pi y),\\
```
```suggestion
   l(x, y) &= \tfrac{1}{2} \pi^2 \cos(\tfrac{1}{2} \pi x) \cos(\tfrac{1}{2} \pi y),\\
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+   \begin{align}
+   l(x, y) &= \tfrac{1}{2} \pi^2 cos(\tfrac{1}{2} \pi x) cos(\tfrac{1}{2} \pi y),\\
+   \lambda(\mu) &= \theta_0(\mu) \lambda_0 + \theta_1(\mu) \lambda_1,\\
+   \theta_0(\mu) &= 1.1 + sin(\mu_0)\mu_1,\\
+   \theta_1(\mu) &= 1.1 + sin(\mu_1),\\
```
```suggestion
   \theta_0(\mu) &= 1.1 + \sin(\mu_0)\mu_1,\\
   \theta_1(\mu) &= 1.1 + \sin(\mu_1),\\
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+functional to the ``minimize`` function. In case no gradient is given,
+``minimize`` just approximates the gradient with finite differences.
+This is not recommended because the gradient is inexact and the
+computation of finite difference requires even more evalutations of the
```
```suggestion
computation of finite differences requires even more evaluations of the
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+
+We can use a standard RB method to build a surrogate model for the FOM.
+As a result, the solution of the primal equation is no longer expensive
+and the optimization method has the chance to converge faster. For this,
```
You mean "faster" in the sense that each solution to the primal problem is faster, or do you mean that less optimization steps are required for convergence?

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+functional. In the FOM example from above we saw that
+:math:`48 - 9 = 39` evaluations of the model were only due to the
+computation of the finite differences. If the problem is more complex
+and the mesh is finer, this can lead us into a serious waste of
```
```suggestion
and the mesh is finer, this can lead to a serious waste of
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+Breaking the traditional offline/online splitting: enrich along the path of optimization
+----------------------------------------------------------------------------------------
+
+We already figured that the main drawback for using RB methods in the
```
```suggestion
We already figured out that the main drawback for using RB methods in the
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+space. Thinking about this issue again, it is important to notice that
+we are solving an optimization problem which will eventually converge to
+a certain parameter. Thus, it only matters that the surrogate is good in
+this particular reason as long as we are able to arrive at it. This
```
```suggestion
this particular region as long as we are able to arrive at it. This
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+we are solving an optimization problem which will eventually converge to
+a certain parameter. Thus, it only matters that the surrogate is good in
+this particular reason as long as we are able to arrive at it. This
+gives hope that there must exists a more efficient way of using RB
```
```suggestion
gives hope that there must exist a more efficient way of using RB
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+a certain parameter. Thus, it only matters that the surrogate is good in
+this particular reason as long as we are able to arrive at it. This
+gives hope that there must exists a more efficient way of using RB
+methods without trying to approximate the whole parameter space.
```
```suggestion
methods without trying to approximate the FOM across the whole parameter space.
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+    
+
+
+The computational time looks at least better than the FOM optimization
+and we are very close to the reference parameter saved some
+computational time. But we are following the exact same path than the
```
I do not understand the last part of the sentence and I am not sure about what you want to actually say here.

TiKeil replied:
For every optimization there is one enrichment step which requires to solve the FOM. I added this. 
Does it now make sense to you?

HenKlei replied:
I think we are talking about different sentences, could that be? I do not understand the last part of 

> The computational time looks at least better than the FOM optimization and we are very close to the reference parameter saved some computational time.

Maybe it is just a language issue or I am just unable to understand what you mean here.

TiKeil replied:
Ah, yes wrong sentence. 
No you are absolutely right, the "saved some computational time" part is completely obsolete, thanks.

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+-----------------------------------
+
+This makes us think about another idea where we only enrich if it is
+neccessary. For example it could be that the model is already good at
```
```suggestion
necessary. For example it could be the case that the model is already good at
```

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,891 @@
+reducing a FOM for an optimization problem. One main result from this
+was that standard RB methods can help to reduce the computational time.
+Thus, standard RB methods are especially of interest if an optimization
+problem might need to solved multiple times.
```
```suggestion
problem might need to be solved multiple times.
```

HenKlei reviewed (commented):
Very nice tutorial! Only some minor remarks from my side...

TiKeil commented:
> Very nice tutorial! Only some minor remarks from my side...

Thanks for the very nice review. I already included all comments.

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,889 @@
+        if (result.status != 0):
+            print('\n failed!')
+        else:
+            print('\n succeded!')
```
```suggestion
            print('\n succeeded!')
```

HenKlei commented on src/pymordemos/linear_optimization.py:
```diff
@@ -0,0 +1,151 @@
+    if (result.status != 0):
+        print('\n failed!')
+    else:
+        print('\n succeded!')
```
```suggestion
        print('\n succeeded!')
```

TiKeil replied:
Will be fixed in #1110

HenKlei replied:
Oh yeah, sorry, I did not recognize that this was not the `.rst` file anymore.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+not change the interpretation of the methods that are discussed below.
+It rather aggravates the result.
+
+Before we start the first optimization method, we define helpful
```
```suggestion
Before we discuss the first optimization method, we define helpful
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+        QoI = function(mu)
+        data['num_evals'] += 1
+        # we need to make sure to copy the data, since the added mu will be changed inplace by minimize afterwards
+        data['evaluation_points'].append([fom.parameters.parse(mu)['diffusion'][:][0], 
+                                          fom.parameters.parse(mu)['diffusion'][:][1]])
```
```suggestion
        data['evaluation_points'].append(fom.parameters.parse(mu).to_numpy())
```
should be sufficient?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+changing the parameter functional :math:`\theta_{\mathcal{J}}` can
+already result in a very different non-convex output functional.
+
+In order to record some data in the optimization, we also define two
```
```suggestion
In order to record some data during the optimization, we also define two
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+Optimizing with the FOM using finite differences
+------------------------------------------------
+
+There exist plenty optimization methods and this tutorial is not meant
```
```suggestion
There exist plenty optimization methods, and this tutorial is not meant
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+                print('  absolute error w.r.t. reference solution: {:.2e}'.format(np.linalg.norm(result.x-reference_mu)))
+            print('  num iterations:     {}'.format(result.nit))
+            print('  num function calls: {}'.format(data['num_evals']))
+            print('  time: {:.5f} seconds'.format(data['time']))
```
maybe align the printed values?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+
+
+Taking a look at the result, we see that the optimizer needs :math:`9`
+iterations to converge, but actually needs :math:`48` evaluations of the
```
In the documentatio generated by CI we have 36 function calls ...

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+
+.. jupyter-execute::
+
+    from pymor.algorithms.greedy import rb_greedy
+    from pymor.reductors.coercive import CoerciveRBReductor
+    
+    from pymor.parameters.functionals import MinThetaParameterFunctional
```
```suggestion
    from pymor.algorithms.greedy import rb_greedy
    from pymor.parameters.functionals import MinThetaParameterFunctional    
    from pymor.reductors.coercive import CoerciveRBReductor
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+    coercivity_estimator = MinThetaParameterFunctional(fom.operator.coefficients, mu_bar)
+    
+    training_set = parameter_space.sample_uniformly(25)
+    training_set_simple = [mu['diffusion'] for mu in training_set]
```
```suggestion
```

Unused?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+    
+    coercivity_estimator = MinThetaParameterFunctional(fom.operator.coefficients, mu_bar)
+    
+    training_set = parameter_space.sample_uniformly(25)
```
Move to code block where greedy training is performed.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+
+
+
+Using MOR for PDE-constrained optimization goes beyond the classical
+online efficiency of RB methods. It is not meaningful to ignore the
```
That sentence is somewhat nebulous.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+optimization methods would already converge before the surrogate model
+is even ready. Thus, the RB optimization methods (at least for only one
+configuration) aims for overall efficiency which includes offline and
+online time. Of course, this effect aggravates if the parameter space is
+high dimensional.
```
Might not be obvious to the reader.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+online time. Of course, this effect aggravates if the parameter space is
+high dimensional.
+
+In order to decrease the offline time we realize that we do not require
+a perfect surrogate model in the sense that a low error tolerance for
+the :func:`~pymor.algorithms.greedy.rb_greedy` already suffices to converge to the same minimum. In
```
How do we realize that?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+    print('RB system is of size {}x{}'.format(num_RB_greedy_extensions, num_RB_greedy_extensions))
+    print('maximum estimated model reduction error over training set: {}'.format(RB_greedy_errors[-1]))
+
+As we can see the greedy already stops after :math:`3` basis functions.
```
```suggestion
As we can see that the greedy algorithm already stops after :math:`3` basis functions.
```

I really don't understand why people like to use 'greedy' as a proper noun.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+A major issue of using finite differences for computing the gradient of
+the objective functional is the number of evaluations of the objective
+functional. In the FOM example from above we saw that
+:math:`48 - 9 = 39` evaluations of the model were only due to the
```
See above.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+      =   \partial_{\mu_i} J(u_{\mu}, \mu) + J(d_{\mu_i} u_{\mu}, \mu)
+   \end{align}
+
+where :math:`d` means total derivative and :math:`\partial` means
+partial derivative. Thus, we need to compute the derivative of the
```
```suggestion
Thus, we need to compute the derivative of the
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+    \label{sens} \tag{2}
+   a_\mu(d_{\mu_i} u_{\mu}, v) = \partial_{\mu_i} r_\mu^{\text{pr}}(u_{\mu})[v] \qquad \qquad \forall v \in V
+
+where :math:`r_\mu^{\text{pr}}` denotes the residual of the primal
+equation. A major issue of this approach is that the computation of the
```
Define 'primal residual'.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+where :math:`r_\mu^{\text{pr}}` denotes the residual of the primal
+equation. A major issue of this approach is that the computation of the
+full gradient requires :math:`P` solutions of :math:`\eqref{sens}`.
+Especially for high dimensional parameter spaces, we can instead use the
```
```suggestion
Especially for high dimensional parameter spaces, we can instead use an
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+.. math::
+
+    \label{dual} \tag{3}
+   a_\mu(v, p_\mu) = \partial_u J(u_\mu, \mu)[v]
```
```suggestion
   a_\mu(v, p_\mu) = \partial_u J(u, \mu)[v]
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+Note that in our case, we then have
+:math:`\mathcal{L}(u_{\mu}, \mu, p_{\mu}) = J(u, \mu)` because the
+residual term :math:`r_\mu^{\text{pr}}(u_{\mu}, p_{\mu})` vanishes. By
+using the dual problem, we can then derive the gradient of the objective
```
```suggestion
using the solution of the dual problem, we can then derive the gradient of the objective
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+----------------------------------
+
+We can easily include a function to compute the gradient to :func:`~scipy.optimize.minimize`.
+For using the adjoint approach we have to explicitly enable the ``use_adjoint`` argument.
```
I thought it was enabled by default?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+For using the adjoint approach we have to explicitly enable the ``use_adjoint`` argument.
+Note that using the (more general) default implementation ``use_adjoint=False`` results
+in the exact same gradient but lacks computational speed.
+Moreover, the function ``output_d_mu`` returns a dict w.r.t. the parameter space as default.
```
```suggestion
Moreover, the function ``output_d_mu`` returns a dict w.r.t. the parameters as default.
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+The online phase is even slightly faster than before but the offline
+phase is obviously still the same as before. We also conclude that the
+ROM model eventually gives less speedup by using a better optimization
+method for the FOM and ROM. This is something that is a common issue for
+MOR methods in the context of optimization.
```
I would remove the last sentence. Sounds a bit like 'don't use better optimization methods'.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+We already figured out that the main drawback for using RB methods in the
+context of optimization is the expensive offline time to build the
+surrogate model. In the example above, we overcame this issue by
+choosing a very high tolerance ``atol``. As a result, we can not be sure
```
```suggestion
choosing a very high tolerance ``atol``. As a result, we cannot be sure
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+whole parameter space.
+
+One possible way for advanced RB methods is a reduction along the path
+of optimization. The idea is, that we start with an empty basis and only
```
```suggestion
of optimization. The idea is that we start with an empty basis and only
```

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+        QoI = rom.output(mu)
+        data['num_evals'] += 1
+        # we need to make sure to copy the data, since the added mu will be changed inplace by minimize afterwards
+        data['evaluation_points'].append([fom.parameters.parse(mu)['diffusion'][:][0], 
+                                          fom.parameters.parse(mu)['diffusion'][:][1]])
```
See above.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+                                          fom.parameters.parse(mu)['diffusion'][:][1]])
+        data['evaluations'].append(QoI[0])
+        opt_dict['opt_rom'] = rom 
+        return QoI
```
Can't you call the `record_results` function from above?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+        opt_rom = opt_dict['opt_rom']
+        return opt_rom.output_d_mu(opt_rom.parameters.parse(mu), return_array=True, use_adjoint=True)
+
+.. jupyter-execute::
```
This seems to be concatenated in the output without any space inbetween. And some text between the cells?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+But we are following the exact same path than the
+FOM and thus we need to solve the FOM model as often as before
+(due to the enrichments). The only time that we safe is the one
+for the dual solution which we compute with the ROM instead.
```
Gradients?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+the next iteration, which we can easily check by evaluating the standard
+error estimator which is also used in the greedy algorithm. In the next
+example we will implement this adaptive way of enriching and set a
+tolerance which is equal to the one that we had in the construction of
+the greedy.
```
How do you construct a greedy?

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+                                          fom.parameters.parse(mu)['diffusion'][:][1]])
+        data['evaluations'].append(QoI[0])
+        opt_dict['opt_rom'] = opt_rom 
+        return QoI
```
See above.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+    assert opt_along_path_minimization_data['enrichments'] == 9
+    assert opt_along_path_adaptively_minimization_data['enrichments'] == 4
+
+
```
This still depends on the magic tolerance for enriching the model. Probably an easy way around this is to do one optimization step for the FOM after converging. If this changes anything, the ROM tolerance was too large. Probably too much work for the tutorial, but one might at least discuss the possibility.

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+adaptive RB methods still reduce the computational demand of the
+optimization method.
+
+For more advanced methods and problems on this topic, we refer to
+Trust-Region methods, quadratic objective functionals, inverse problems
+or higher order optimization methods. A Trust-Region method for
+quadratic objective functionals with a non-conforming RB approach has
+been considered in `this (arXiv)
+paper <https://arxiv.org/abs/2006.09297>`__, where pyMOR has been used
+for the MOR part. You can see the whole code and all numerical results
+in `this
+project <https://github.com/TiKeil/NCD-corrected-TR-RB-approach-for-pde-opt>`__.
```
Only citing your own preprint regarding RB methods for optimization sounds like a bit too much of self-advertisement. Please also reference some published standard literature as well.

sdrave reviewed (changes requested):
In general, this is a very nice tutorial! But maybe try to give a more direct and positive answer to: 'I have a PDE-constrained optimization problem, and I want to use pyMOR for speedup. What should I try as a first step?'  Reading through the very long text, at the end you might get the impression that only an open problem has been presented. You should have a feeling of achievement.

TiKeil commented:
Thanks @sdrave for the very nice and detailed review. I have addressed all your points except the citations.

I think, the text in the end reads way more positive now. I will see how I can improve this more but it should be okay already.

Removing Draft mode.

renefritze commented:
> Thanks @sdrave for the very nice and detailed review. I have addressed all your points except the citations.
> 
> I think, the text in the end reads way more positive now. I will see how I can improve this more but it should be okay already.
>
 
[Looking good](https://docs.pymor.org/github-push-tutorial-optimization/tutorial_optimization.html) :+1: 
If you could get your last tweaks in by tomorrow evening, that would be great. 

Also, can you do me a favor and also fix a typo in the mybinder box?

```Please note that starting the notebook server make take a couple of minutes.```

TiKeil commented:
> > Thanks @sdrave for the very nice and detailed review. I have addressed all your points except the citations.
> > I think, the text in the end reads way more positive now. I will see how I can improve this more but it should be okay already.
> 
> [Looking good](https://docs.pymor.org/github-push-tutorial-optimization/tutorial_optimization.html) +1
> If you could get your last tweaks in by tomorrow evening, that would be great.
> 
> Also, can you do me a favor and also fix a typo in the mybinder box?
> 
> `Please note that starting the notebook server make take a couple of minutes.`

Thanks, I have fixed the typo and increased the timeout limit. 

Furthermore I have added my last changes. Thus, this can be merged if there are no other objections.

HenKlei commented:
You could also add the references to [the bibliography file](https://github.com/pymor/pymor/blob/master/docs/source/bibliography.rst).

renefritze reviewed (approved):
Greenlight for CI again

HenKlei commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+optimization problems, we refer to 
+`YM13 <https://epubs.siam.org/doi/abs/10.1137/120869171>`__, 
+`QGVW17 <https://epubs.siam.org/doi/abs/10.1137/16M1081981>`__ and 
+`KMSOV <https://arxiv.org/abs/2006.09297>`__ where for the latter, PyMOR
```
```suggestion
`KMSOV <https://arxiv.org/abs/2006.09297>`__ where for the latter, pyMOR
```

TiKeil commented:
> You could also add the references to [the bibliography file](https://github.com/pymor/pymor/blob/master/docs/source/bibliography.rst).

I intentionally did not do that because these references do not have anything to do with algorithms or implementational features in
PyMOR. Meaning, we do not provide any sort of content from these references. They are rather meant as `further reading` for the interested tutorial reader.

HenKlei commented:
> > You could also add the references to [the bibliography file](https://github.com/pymor/pymor/blob/master/docs/source/bibliography.rst).
> 
> I intentionally did not do that because these references do not have anything to do with algorithms or implementational features in
> pyMOR. Meaning, we do not provide any sort of content from these references. They are rather meant as "further reading" for the interested tutorial reader.

Ah okay, sounds reasonable.

ftalbrecht commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+just approximates the gradient with finite differences.
+This is not recommended because the gradient is inexact and the
+computation of finite differences requires even more evaluations of the
+primal equation. We anyway start with this approach.
```
```suggestion
primal equation. Here, we use this approach for a simple demonstration.
```

ftalbrecht commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+   a_\mu(v, p_\mu) = \partial_u J(u_\mu, \mu)[v]
+   = J(v, \mu)
+
+Note that in our case, we then have
```
Perhaps mention quickly why it vanishes here (spaces and so on).

TiKeil replied:
I wrote half a sentence more

ftalbrecht commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+ROM model eventually gives less speedup by using a better optimization
+method for the FOM and ROM. 
+
+Breaking the traditional offline/online splitting: enrich along the path of optimization
```
```suggestion
Beyond the traditional offline/online splitting: enrich along the path of optimization
```

ftalbrecht commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+We already figured out that the main drawback for using RB methods in the
+context of optimization is the expensive offline time to build the
+surrogate model. In the example above, we overcame this issue by
+choosing a very high tolerance ``atol``. As a result, we cannot be sure
```
```suggestion
choosing a large value for ``atol``. As a result, we cannot be sure
```

ftalbrecht commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+Adaptively enriching along the path
+-----------------------------------
+
+This makes us think about another idea where we only enrich if it is
```
```suggestion
This gives rise to another idea where we only enrich if it is
```

ftalbrecht commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+the next iteration, which we can easily check by evaluating the standard
+error estimator which is also used in the greedy algorithm. In the next
+example we will implement this adaptive way of enriching and set a
+tolerance which is equal to the one that we had in the as error tolerance
```
```suggestion
tolerance which is equal to the one that we had as error tolerance
```

ftalbrecht commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,894 @@
+Conclusion and some general words about MOR methods for optimization
+--------------------------------------------------------------------
+
+In this tutorial we have seen how PyMOR can be used to speedup the optimizer
```
```suggestion
In this tutorial we have seen how pyMOR can be used to speedup the optimizer
```

ftalbrecht reviewed (changes requested):
After fixing everything, I suggest to
- go through the intro once and unify the number of whitespace lines between the code and the text (say one before/after code blocks and two before headlines) and
- check the indentation of the code examples, in particular the creation of the data dicts.

TiKeil commented:
> After fixing everything, I suggest to
> 
>     * go through the intro once and unify the number of whitespace lines between the code and the text (say one before/after code blocks and two before headlines) and
> 
>     * check the indentation of the code examples, in particular the creation of the data dicts.

Thanks @ftalbrecht, I have added the files accordingly. Some of your remarks were outdated already such that I could not apply all requested changes.

TiKeil commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+
+In this tutorial, we use a simple linear scalar valued objective functional
+and an elliptic primal equation to compare different approaches that solve
+(:math:`\hat{P}`). For a more advanced problem class, we refer to 
```
I do not mention the reference here anymore due to the comment of @sdrave because I do not keep the referenced github project up to date with the pyMOR releases.

TiKeil commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+.. jupyter-execute::
+
+    output_functional = fom.rhs.H * theta_J 
+    fom = fom.with_(output_functional=output_functional)
```
You are right, I removed the block.

TiKeil commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,900 @@
+integrals over :math:`u_\mu`, multiplied by an arbitrary |Function|,
+are supported. These outputs can also be handled by the discretizer. In
+our case, we make use of the :math:`L^2` output and multiply it by the
+term :math:`\theta_\mu l_\mu`.
```
This text block was removed already

sdrave commented on docs/source/tutorial_optimization.rst:
```diff
@@ -0,0 +1,842 @@
+the path of optimization or (even better) only enrich
+the model if the standard error estimator goes above a certain tolerance.
+
+A main drawback of the content in this tutorial was that the choice of
+the tolerance ``atol`` that has been used to build the RB spaces
+cannot be known a priorily. This shows the need for
+certified and robust reduced methods. 
+
+For some standard literature for faster and robust optimization tools we refer to
+`CGT00 <https://epubs.siam.org/doi/book/10.1137/1.9780898719857?mobileUi=0>`__ and
+`NW06 <https://link.springer.com/book/10.1007/978-0-387-40065-5>`__.
+
+For recent research on using trust-region methods for MOR of PDE-constrained
+optimization problems, we refer to 
+`YM13 <https://epubs.siam.org/doi/abs/10.1137/120869171>`__, 
+`QGVW17 <https://epubs.siam.org/doi/abs/10.1137/16M1081981>`__ and 
+`KMSOV <https://arxiv.org/abs/2006.09297>`__ where for the latter, pyMOR
+has been used for the numerical experiments.
```
```suggestion
In this tutorial we have only covered a few basic approaches to combine model
reduction with optimization.
For faster and more robust optimization algorithms we refer to the textbooks
`CGT00 <https://epubs.siam.org/doi/book/10.1137/1.9780898719857>`__ and
`NW06 <https://link.springer.com/book/10.1007/978-0-387-40065-5>`__.
For recent research on combining trust-region methods with model reduction for
PDE-constrained optimization problems we refer to 
`YM13 <https://epubs.siam.org/doi/abs/10.1137/120869171>`__, 
`QGVW17 <https://epubs.siam.org/doi/abs/10.1137/16M1081981>`__ and 
`KMSOV <https://arxiv.org/abs/2006.09297>`__ where for the latter a pyMOR
implementation is available as supplementary material.
```
